pam/gdm/conversation_test.go

```diff
@@ -99,3 +129,569 @@ func TestSendToGdm(t *testing.T) {
 		})
 	}
 }
+
+func TestDataSend(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(pam_test.MaybeDoLeakCheck)
+	testCases := map[string]struct {
+		value *Data
+
+		wantReturn []byte
+		wantError  error
+	}{
+		"nil return": {
+			value: &Data{
+				Type: DataType_event,
+				Event: &EventData{
+					Type: EventType_brokerSelected,
+					Data: &EventData_BrokerSelected{},
+				},
+			},
+
+			wantReturn: []byte("null"),
+		},
+
+		"version data": {
+			value: &Data{
+				Type:  DataType_hello,
+				Hello: &HelloData{Version: 12345},
+			},
+			wantReturn: []byte(`"hello gdm!"`),
+		},
+
+		// Error cases
+		"Error on empty data": {
+			value:      &Data{},
+			wantReturn: nil,
+			wantError:  errors.New("expected type unknownType"),
+		},
+
+		"Error on missing data return": {
+			value: &Data{
+				Type: DataType_event,
+				Event: &EventData{
+					Type: EventType_brokerSelected,
+					Data: nil,
+				},
+			},
+
+			wantError: errors.New("missing event data"),
+		},
+
+		"Error on wrong data": {
+			value: &Data{
+				Type:    DataType_event,
+				Request: &RequestData{},
+			},
+			wantError: errors.New("missing event type"),
+		},
+	}
+
+	for name, tc := range testCases {
```

Comment:
We also talked about this one between the test cases and the loop...

Reply:
Yeah, most of this code was very old so once split I forgot to address some of the comments from other PR's.